test/py/libvfio_user.py

```diff
@@ -0,0 +1,167 @@
+#  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+#  DAMAGE.
+#
+
+#
+# Note that we don't use enum here, as class.value is a little verbose
+#
+
+from types import SimpleNamespace
+import ctypes as c
+import json
+import os
+import pathlib
+import socket
+import struct
+import syslog
+
+SOCK_PATH = b"/tmp/vfio-user.sock.%d" % os.getpid()
+
+VFU_TRANS_SOCK = 0
+LIBVFIO_USER_FLAG_ATTACH_NB = (1 << 0)
+VFU_DEV_TYPE_PCI = 0
+
+LIBVFIO_USER_MAJOR = 0
+LIBVFIO_USER_MINOR = 1
+
+VFIO_USER_CLIENT_MAX_FDS_LIMIT = 1024
+
+SERVER_MAX_FDS = 8
+
+SERVER_MAX_MSG_SIZE = 65536
+
+# enum vfio_user_command
+VFIO_USER_VERSION	    			= 1
+VFIO_USER_DMA_MAP	    			= 2
+VFIO_USER_DMA_UNMAP	    			= 3
+VFIO_USER_DEVICE_GET_INFO			= 4
+VFIO_USER_DEVICE_GET_REGION_INFO	= 5
+VFIO_USER_DEVICE_GET_IRQ_INFO		= 6
+VFIO_USER_DEVICE_SET_IRQS   		= 7
+VFIO_USER_REGION_READ				= 8
+VFIO_USER_REGION_WRITE				= 9
+VFIO_USER_DMA_READ			    	= 10
+VFIO_USER_DMA_WRITE   				= 11
+VFIO_USER_VM_INTERRUPT				= 12
+VFIO_USER_DEVICE_RESET				= 13
+VFIO_USER_DIRTY_PAGES				= 14
+VFIO_USER_MAX			        	= 15
+
+VFIO_USER_F_TYPE_COMMAND = 0
+VFIO_USER_F_TYPE_REPLY = 1
+
+SIZEOF_VFIO_USER_HEADER = 16
+
+topdir = os.path.realpath(os.path.dirname(__file__) + "/../..")
+build_type = os.getenv("BUILD_TYPE", default="dbg")
+libname = "%s/build/%s/lib/libvfio-user.so" % (topdir, build_type)
+lib = c.CDLL(libname, use_errno=True)
+lib.vfu_create_ctx.argtypes = (c.c_int, c.c_char_p, c.c_int,
+                               c.c_void_p, c.c_int)
+lib.vfu_create_ctx.restype = (c.c_void_p)
+lib.vfu_setup_log.argtypes = (c.c_void_p, c.c_void_p, c.c_int)
+lib.vfu_realize_ctx.argtypes = (c.c_void_p,)
+lib.vfu_attach_ctx.argtypes = (c.c_void_p,)
+lib.vfu_run_ctx.argtypes = (c.c_void_p,)
+lib.vfu_destroy_ctx.argtypes = (c.c_void_p,)
+
+msg_id = 1
+
+#
+# Util functions
+#
+
+def connect_sock():
+    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+    sock.connect(SOCK_PATH)
+    return sock
+
+def get_reply(sock, expect=0):
+    buf = sock.recv(4096)
+    (msg_id, cmd, msg_size, flags, errno) = struct.unpack("HHIII", buf[0:16])
+    assert errno == expect
+    msg_size -= 16
+    return buf[16:]
+
+#
+# Parse JSON into an object with attributes (instead of using a dict)
+#
+def parse_json(json_str):
+    return json.loads(json_str, object_hook=lambda d: SimpleNamespace(**d))
+
+
+#
+# Library wrappers
+#
+
+@c.CFUNCTYPE(None, c.c_void_p, c.c_int, c.c_char_p)
+def log(ctx, level, msg):
+    print(msg.decode("utf-8"))
+
+def vfio_user_header(cmd, size, no_reply=False, error=False, error_no=0):
+    global msg_id
+
+    buf = struct.pack("HHIII", msg_id, cmd, SIZEOF_VFIO_USER_HEADER + size,
+                      0, error_no)
+
+    msg_id += 1
+
+    return buf
+
+def vfu_create_ctx(trans=VFU_TRANS_SOCK, sock_path=SOCK_PATH, flags=0,
+                   private=None, dev_type=VFU_DEV_TYPE_PCI):
+    if os.path.exists(sock_path):
+        os.remove(sock_path)
```

Comment:
nit: just attempt to remove the path and handle the `ENOENT` exception?

Reply:
I think it's clearer as is?

Reply:
In theory this is vulnerable to TOCTOU bugs, but we shouldn't have to deal with things like that in the testing framework.